### PR TITLE
Hexagonal to cuboid functionality

### PR DESCRIPTION
**Pull request by mlund**
Adds function to convert hexagonal prism to cuboid and apply it in `savestate` for PQR output with the optional keyword `convert_hexagon`.

### CODE REVIEW

**Review by rc83 — CHANGES_REQUESTED** [review]
The algorithm is surprisingly simple, isn’t it? Beside minor comments, I have one big no-no regarding references to a unique pointer. ¶ The more I think about the hexagonal geometry, the more I am persuaded that Faunus should internally work with a rhombic geometry with transformation from and to hexagonal one on input and output, respectively.

File: src/geometry.cpp [hunk 1]
```diff
@@ -714,6 +735,8 @@ Chameleon::Chameleon(const Chameleon &geo)
     : GeometryBase(geo), len(geo.len), len_half(geo.len_half), len_inv(geo.len_inv),
       geometry(geo.geometry != nullptr ? geo.geometry->clone() : nullptr), _type(geo._type), _name(geo._name) {}
 
+std::unique_ptr<GeometryImplementation> &Chameleon::getGeometryImplementation() { return geometry; }
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/geometry.cpp:738]: fixed!

File: src/analysis.cpp [hunk 2]
```diff
@@ -195,13 +195,28 @@ SaveState::SaveState(json j, Space &spc) {
...
     if (auto suffix = filename.substr(filename.find_last_of(".") + 1); suffix == "aam") {
         writeFunc = [&](auto &file) { FormatAAM::save(file, spc.p); };
     } else if (suffix == "gro") {
         writeFunc = [&](auto &file) { FormatGRO::save(file, spc); };
     } else if (suffix == "pqr") {
-        writeFunc = [&](auto &file) { FormatPQR::save(file, spc.groups, spc.geo.getLength()); };
+        writeFunc = [&](auto &file) {
+            if (convert_hexagonal_prism_to_cuboid) {
+                auto hexagonal_prism =
+                    dynamic_cast<Geometry::HexagonalPrism *>(spc.geo.getGeometryImplementation().get());
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/analysis.cpp:208]: fixed

File: src/analysis.cpp [hunk 3]
```diff
@@ -195,13 +195,28 @@ SaveState::SaveState(json j, Space &spc) {
...
+                if (hexagonal_prism) {
+                    faunus_logger->debug("creating cuboidal PQR from hexagonal prism");
+                    const auto &[cuboid, particles] = Geometry::HexagonalPrismToCuboid(*hexagonal_prism, spc.p);
+                    FormatPQR::save(file, particles, cuboid.getLength());
+                } else {
+                    throw std::runtime_error("hexagonal prism required for `convert_to_hexagon`");
+                }
+            } else {
+                FormatPQR::save(file, spc.groups, spc.geo.getLength());
+            }
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/analysis.cpp:218]: I agree, that's the future and more economic way to go!

File: src/geometry.cpp [hunk 4]
```diff
@@ -591,6 +595,23 @@ ParticleVector mapParticlesOnSphere(const ParticleVector &source) {
     return destination;
 }
 
+std::pair<Cuboid, ParticleVector> HexagonalPrismToCuboid(const HexagonalPrism &hexagon,
+                                                         const ParticleVector &particles) {
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/geometry.cpp:599]: Yes, a range would be neat but we still need to create and return the Cuboid object.

File: src/geometry.cpp [hunk 5]
```diff
@@ -591,6 +595,23 @@ ParticleVector mapParticlesOnSphere(const ParticleVector &source) {
...
+    std::transform(particles.begin(), particles.end(), std::back_inserter(cuboid_particles), [&](auto particle) {
+        particle.pos.x() += hexagon.innerRadius() * (particle.pos.x() > 0.0 ? -1.0 : 1.0);
+        particle.pos.y() += hexagon.outerRadius() * (particle.pos.y() > 0.0 ? -1.0 : 1.0) * 1.5;
+        assert(cuboid.collision(particle.pos) == false);
+        return particle;
+    }); // add the four corners; i.e. one extra, split hexagon
+    assert(std::fabs(cuboid.getVolume() - 2.0 * hexagon.getVolume()) <= pc::epsilon_dbl);
+    return {cuboid, cuboid_particles};
+}
+
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/geometry.cpp:614]: fixed!

**Pull request merged**